File: instagram-handler/services/temp_email.py
import random
import re
import string
import time
from typing import Any

import requests

CODE_PATTERN = re.compile(r"\b(\d{6})\b")

PROVIDERS = [
  "mailtm",
  "1secmail",
]


class TempEmailError(Exception):
    pass


class TempEmail:
    def __init__(self, address: str | None = None, provider: str | None = None):
        self.provider = provider or "mailtm"
        self.token = ""
        self.password = ""
        if address:
            self.address = address
            self.login, self.domain = address.split("@", 1)
        else:
            self._create_mailbox()

    def _create_mailbox(self) -> None:
        errors = []
        providers = [self.provider] + [p for p in PROVIDERS if p != self.provider]
        for name in providers:
            try:
                if name == "mailtm":
                    self._create_mailtm()
                elif name == "1secmail":
                    self._create_1secmail()
                self.provider = name
                return
            except Exception as exc:
                errors.append(f"{name}: {exc}")
        raise TempEmailError("All temp email providers failed — " + "; ".join(errors))

    def _create_mailtm(self) -> None:
        domains_resp = requests.get("https://api.mail.tm/domains", timeout=15)
        domains_resp.raise_for_status()
        domains = [d["domain"] for d in domains_resp.json().get("hydra:member", []) if d.get("isActive")]
        if not domains:
            raise TempEmailError("No mail.tm domains available")
        self.domain = random.choice(domains)
        self.login = "ig" + "".join(random.choices(string.ascii_lowercase + string.digits, k=10))
        self.address = f"{self.login}@{self.domain}"
        self.password = "".join(random.choices(string.ascii_letters + string.digits, k=16))
        acc_resp = requests.post(
            "https://api.mail.tm/accounts",
            json={"address": self.address, "password": self.password},
            timeout=15,
        )
        acc_resp.raise_for_status()
        token_resp = requests.post(
            "https://api.mail.tm/token",
            json={"address": self.address, "password": self.password},
            timeout=15,
        )
        token_resp.raise_for_status()
        self.token = token_resp.json().get("token", "")

    def _create_1secmail(self) -> None:
        resp = requests.get(
            "https://www.1secmail.com/api/v1/?action=genRandomMailbox&count=1",
            timeout=15,
            headers={"User-Agent": "Mozilla/5.0"},
        )
        resp.raise_for_status()
        data = resp.json()
        if not data:
            raise TempEmailError("Failed to generate 1secmail address")
        self.address = data[0]
        self.login, self.domain = self.address.split("@", 1)

    def get_messages(self) -> list[dict[str, Any]]:
        if self.provider == "mailtm":
            resp = requests.get(
                "https://api.mail.tm/messages",
                headers={"Authorization": f"Bearer {self.token}"},
                timeout=15,
            )
            resp.raise_for_status()
            return resp.json().get("hydra:member", [])
        resp = requests.get(
            f"https://www.1secmail.com/api/v1/?action=getMessages&login={self.login}&domain={self.domain}",
            timeout=15,
            headers={"User-Agent": "Mozilla/5.0"},
        )
        resp.raise_for_status()
        return resp.json()

    def read_message(self, msg_id: str | int) -> dict[str, Any]:
        if self.provider == "mailtm":
            resp = requests.get(
                f"https://api.mail.tm/messages/{msg_id}",
                headers={"Authorization": f"Bearer {self.token}"},
                timeout=15,
            )
            resp.raise_for_status()
            data = resp.json()
            return {
                "subject": data.get("subject", ""),
                "textBody": data.get("text", ""),
                "htmlBody": data.get("html", ""),
            }
        resp = requests.get(
            f"https://www.1secmail.com/api/v1/?action=readMessage&login={self.login}&domain={self.domain}&id={msg_id}",
            timeout=15,
            headers={"User-Agent": "Mozilla/5.0"},
        )
        resp.raise_for_status()
        return resp.json()

    def extract_code(self, text: str) -> str | None:
        match = CODE_PATTERN.search(text or "")
        return match.group(1) if match else None

    def wait_for_code(self, timeout: int = 120, poll_interval: int = 5) -> str:
        seen: set[str] = set()
        deadline = time.time() + timeout
        while time.time() < deadline:
            for msg in self.get_messages():
                msg_id = str(msg.get("id", ""))
                if not msg_id or msg_id in seen:
                    continue
                seen.add(msg_id)
                if self.provider == "mailtm":
                    body = f"{msg.get('subject', '')} {msg.get('intro', '')}"
                    full = self.read_message(msg_id)
                    body += f" {full.get('textBody', '')} {full.get('htmlBody', '')}"
                else:
                    full = self.read_message(msg_id)
                    body = f"{full.get('subject', '')} {full.get('textBody', '')} {full.get('htmlBody', '')}"
                code = self.extract_code(body)
                if code:
                    return code
            time.sleep(poll_interval)
        raise TempEmailError(f"Verification code not received within {timeout}s")

File: www/app/src/TempEmail.php
<?php

declare(strict_types=1);

final class TempEmail
{
    public string $address;
    public string $login;
    public string $domain;
    public string $provider = 'mailtm';
    private string $token = '';
    private string $password = '';

    public function __construct(?string $address = null)
    {
        if ($address) {
            $this->address = $address;
            [$this->login, $this->domain] = explode('@', $address, 2);
        } else {
            $this->createMailbox();
        }
    }

    private function createMailbox(): void
    {
        $errors = [];
        foreach (['mailtm', '1secmail'] as $provider) {
            try {
                if ($provider === 'mailtm') {
                    $this->createMailtm();
                } else {
                    $this->create1SecMail();
                }
                $this->provider = $provider;
                return;
            } catch (Throwable $e) {
                $errors[] = "$provider: " . $e->getMessage();
            }
        }
        throw new RuntimeException('All temp email providers failed — ' . implode('; ', $errors));
    }

    private function createMailtm(): void
    {
        $domains = $this->httpJson('GET', 'https://api.mail.tm/domains');
        $list = $domains['hydra:member'] ?? (array_is_list($domains) ? $domains : []);
        $active = array_values(array_filter($list, fn($d) => !empty($d['isActive'])));
        if (!$active) {
            throw new RuntimeException('No mail.tm domains');
        }
        $pick = $active[array_rand($active)];
        $this->domain = $pick['domain'];
        $this->login = 'ig' . bin2hex(random_bytes(5));
        $this->address = $this->login . '@' . $this->domain;
        $this->password = bin2hex(random_bytes(8));
        $this->httpJson('POST', 'https://api.mail.tm/accounts', [
            'address' => $this->address,
            'password' => $this->password,
        ]);
        $token = $this->httpJson('POST', 'https://api.mail.tm/token', [
            'address' => $this->address,
            'password' => $this->password,
        ]);
        $this->token = $token['token'] ?? '';
    }

    private function create1SecMail(): void
    {
        $data = $this->httpJson(
            'GET',
            'https://www.1secmail.com/api/v1/?action=genRandomMailbox&count=1',
            null,
            ['User-Agent: Mozilla/5.0']
        );
        if (empty($data[0])) {
            throw new RuntimeException('1secmail failed');
        }
        $this->address = $data[0];
        [$this->login, $this->domain] = explode('@', $this->address, 2);
    }

    public function waitForCode(int $timeout = 120, int $poll = 5): string
    {
        $seen = [];
        $deadline = time() + $timeout;
        while (time() < $deadline) {
            foreach ($this->getMessages() as $msg) {
                $id = (string) ($msg['id'] ?? '');
                if ($id === '' || isset($seen[$id])) {
                    continue;
                }
                $seen[$id] = true;
                $body = $this->readMessageBody($msg);
                if (preg_match('/\b(\d{6})\b/', $body, $m)) {
                    return $m[1];
                }
            }
            sleep($poll);
        }
        throw new RuntimeException("Verification code not received within {$timeout}s");
    }

    private function getMessages(): array
    {
        if ($this->provider === 'mailtm') {
            $data = $this->httpJson('GET', 'https://api.mail.tm/messages', null, [
                'Authorization: Bearer ' . $this->token,
            ]);
            return $data['hydra:member'] ?? (array_is_list($data) ? $data : []);
        }
        return $this->httpJson(
            'GET',
            "https://www.1secmail.com/api/v1/?action=getMessages&login={$this->login}&domain={$this->domain}",
            null,
            ['User-Agent: Mozilla/5.0']
        );
    }

    private function readMessageBody(array $msg): string
    {
        if ($this->provider === 'mailtm') {
            $id = $msg['id'] ?? '';
            $full = $this->httpJson('GET', "https://api.mail.tm/messages/$id", null, [
                'Authorization: Bearer ' . $this->token,
            ]);
            return ($msg['subject'] ?? '') . ' ' . ($msg['intro'] ?? '') . ' ' . ($full['text'] ?? '') . ' ' . ($full['html'] ?? '');
        }
        $id = $msg['id'] ?? '';
        $full = $this->httpJson(
            'GET',
            "https://www.1secmail.com/api/v1/?action=readMessage&login={$this->login}&domain={$this->domain}&id=$id",
            null,
            ['User-Agent: Mozilla/5.0']
        );
        return ($full['subject'] ?? '') . ' ' . ($full['textBody'] ?? '') . ' ' . ($full['htmlBody'] ?? '');
    }

    private function httpJson(string $method, string $url, ?array $body = null, array $headers = []): array
    {
        $ch = curl_init($url);
        $opts = [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 20,
            CURLOPT_HTTPHEADER => array_merge(['Accept: application/json'], $headers),
        ];
        if ($method === 'POST') {
            $opts[CURLOPT_POST] = true;
            $opts[CURLOPT_HTTPHEADER][] = 'Content-Type: application/json';
            $opts[CURLOPT_POSTFIELDS] = json_encode($body ?? []);
        }
        curl_setopt_array($ch, $opts);
        $resp = curl_exec($ch);
        $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        if ($resp === false) {
            throw new RuntimeException('HTTP error: ' . curl_error($ch));
        }
        curl_close($ch);
        if ($code >= 400) {
            throw new RuntimeException("HTTP $code for $url");
        }
        $data = json_decode($resp, true);
        return is_array($data) ? $data : [];
    }
}
